cmd/server-manager/plugins/events.lua

```diff
@@ -12,20 +16,49 @@ function onEventStart(encodedRaceConfig, encodedServerOpts, encodedEntryList)
     local serverOpts = json.decode(encodedServerOpts)
     local entryList = json.decode(encodedEntryList)
 
+
     -- Uncomment these lines and run the function (start any event) to print out the structure of each object.
     --print("Race Config:", utils.dump(raceConfig))
     --print("Server Options:", utils.dump(serverOpts))
     --print("Entry List:", utils.dump(entryList))
-
+    --print("Countrycodes", utils.dump(countryCodes))
     -- Function block NOTE: this hook BLOCKS, make sure your functions don't loop forever!
-    -- in order to use the weatherAPI you need to get a free API key from https://openweathermap.org/
-    -- uncomment functions to enable them!
-    --raceConfig, serverOpts = weatherAPI(raceConfig, serverOpts, "get-an-api-key-from-https://openweathermap.org/")
+    --Uncomment this line to set Weather API On, don't forget to put you openweathermap API key line 55
+    getTrackInfo(raceConfig, serverOpts)
 
-    -- Encode block, you probably shouldn't touch these either!
+-- Encode block, you probably shouldn't touch these either!
     return json.encode(entryList), json.encode(serverOpts), json.encode(raceConfig)
 end
 
+-- get track location from his ui.json file, set dynamic weather on if a location is found
+-- if not, use weather in custom race config
+function getTrackInfo(raceConfig, serverOpts)
+    local track = raceConfig["Track"]
+    local layout = raceConfig["TrackLayout"]
+    local trackPath = contentPath .. "/tracks/" .. track .. "/ui/" .. layout
+    local encodedTrackJson = utils.jsonOpen(trackPath, "ui_track.json")
+    local trackJson = json.decode(encodedTrackJson)
+    countryFull = trackJson["country"]
+    city = trackJson["city"]
+    -- If Country or City is missing, stop the function and use weather configured in web manager
+    if city == nil or countryFull == nil then
+	print("No location found on track UI file, dynamic weather is OFF")
+	return raceConfig, serverOpts
```

Comment:
Do we need to fail here? I think in this case I would like to at least show people that they can manually set the location, maybe just a commented line or two:

```
-- if you want to manually set the location for tracks without location info uncomment these lines and set the location
-- location = "Manchester,uk"
-- raceConfig, serverOpts = weatherAPI(raceConfig, serverOpts, "get-an-api-key-from-https://openweathermap.org/")

return raceConfig, serverOpts
```

Reply:
I'm thinking to remove it, the return of weatherAPI line 98/99 with error 404 can be sufficient for that, what do you think?

Reply:
I've read through this again, and I think this would be best (in the ``onEventStart`` function)
```
location = getTrackInfo(raceConfig, serverOpts) -- getTrackInfo needs to return location (and you're right, it doesn't modify config or opts so no need to return them).

if location == nil then
    -- manually set location or return options
end

raceConfig, serverOpts = weatherAPI(raceConfig, serverOpts, "get-an-api-key-from-https://openweathermap.org/")
```

then this could just return nil, and you deal with in the main function. It's best to deal with errors as soon as you can rather than letting the program run on when you know it isn't going to work, especially if you're making requests to an API.

Reply:
"if location == nil then" : can't be nil because if city or country is nil at this time =>     	location = city .. "," .. countryCodes[countryFull]
The script will be in error as it can't concatenat string and nil together

Reply:
That's in the else, not this if. In this if set location = nil then return